### PR TITLE
[feathr] Add strong typing with protobuf for redis push and redis get

### PR DESCRIPTION
**Pull request by hangfei**
Add protobuf  so we can have stronger typing for redis push and redis get.

It also allows us to support more types in the future.

### CODE REVIEW

File: feathr_project/feathr/protobuf/featureValue_pb2.py [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
...seFloatArrayH\x00\x42\x13\n\x11\x46\x65\x61tureValueOneOf\" \n\x0c\x42ooleanArray\x12\x10\n\x08\x62ooleans\x18\x01 \x03(\x08\"\x1e\n\x0bStringArray\x12\x0f\n\x07strings\x18\x01 \x03(\t\"\x1e\n\x0b\x44oubleArray\x12\x0f\n\x07\x64oubles\x18\x01 \x03(\x01\"\x1c\n\nFloatArray\x12\x0e\n\x06\x66loats\x18\x01 \x03(\x02\"\x1c\n\x0cIntegerArray\x12\x0c\n\x04ints\x18\x01 \x03(\x05\"\x1a\n\tLongArray\x12\r\n\x05longs\x18\x01 \x03(\x03\"!\n\nBytesArray\x12\x13\n\x0b\x62ytes_array\x18\x01 \x03(\x0c\"6\n\x11SparseStringArray\x12\x10\n\x08integers\x18\x01 \x03(\x05\x12\x0f\n\x07strings\x18\x02 \x03(\t\"5\n\x0fSparseBoolArray\x12\x10\n\x08integers\x18\x01 \x03(\x05\x12\x10\n\x08\x62ooleans\x18\x02 \x03(\x08\"D\n\x12SparseIntegerArray\x12\x16\n\x0eindex_integers\x18\x01 \x03(\x05\x12\x16\n\x0evalue_integers\x18\x02 \x03(\x05\"8\n\x0fSparseLongArray\x12\x16\n\x0eindex_integers\x18\x01 \x03(\x05\x12\r\n\x05longs\x18\x02 \x03(\x03\"6\n\x11SparseDoubleArray\x12\x10\n\x08integers\x18\x01 \x03(\x05\x12\x0f\n\x07\x64oubles\x18\x02 \x03(\x01\"4\n\x10SparseFloatArray\x12\x10\n\x08integers\x18\x01 \x03(\x05\x12\x0e\n\x06\x66loats\x18\x02 \x03(\x02\x42+\n)com.linkedin.feathr.common.types.protobufb\x06proto3')
```
Comment (xiaoyongzhu) [inline feathr_project/feathr/protobuf/featureValue_pb2.py:17]: Add some comments on where this is from (or how to update it)?

File: feathr_project/test/test_azure_synapse_e2e.py [hunk 2]
```diff
@@ -42,6 +46,57 @@ def test_feathr_online_store():
...
+def test_feathr_online_store_non_agg_features():
+    """
+    Test FeathrClient() online_get_features and batch_get can get data correctly.
+    """
+    test_workspace_dir = Path(__file__).parent.resolve() / "test_user_workspace"
+    os.chdir(test_workspace_dir)
+    client = FeathrClient()
+
+
+    # client.materialize_features()
```
Comment (jaymo001) [inline feathr_project/test/test_azure_synapse_e2e.py:58]: Isn't this necessarly?

File: feathr_project/setup.py [hunk 3]
```diff
@@ -34,6 +34,8 @@
         "python-snappy",
         "pyyaml",
         "Jinja2",
+        "google",
+        "google-api-python-client",
```
Comment (xiaoyongzhu) [inline feathr_project/setup.py:38]: You might want to put some versioning here (like larger than a certain version) to make the maintenance easier.

File: feathr_project/test/test_azure_synapse_e2e.py [hunk 4]
```diff
@@ -42,6 +46,57 @@ def test_feathr_online_store():
...
+    # # just assume the job is successful without validating the actual result in Redis. Might need to consolidate
+    # # this part with the test_feathr_online_store test case
+    # client.wait_job_to_finish(timeout_sec=600)
+    res = client.online_get_features('nycTaxiDemoFeature', '111', ['f_gen_trip_distance', 'f_gen_is_long_trip_distance',
```
Comment (jaymo001) [inline feathr_project/test/test_azure_synapse_e2e.py:62]: Where are these 'f1', 'f2', 'f3', 'f4', 'f5', 'f6' features from?  Overall, better description of this test case would be better.

File: feathr_project/test/test_azure_synapse_e2e.py [hunk 5]
```diff
@@ -42,6 +46,57 @@ def test_feathr_online_store():
...
+    assert res[0] != None
+    assert res[1] != None
+    # assert constant features
+    assert res[2] == [10.0, 20.0, 30.0]
+    assert res[3] == ['a', 'b', 'c']
+    assert res[4] == ([1, 2, 3], ['10', '20', '30'])
+    assert res[5] == ([1, 2, 3], [True, False, True])
+    assert res[6] == ([1, 2, 3], [1.0, 2.0, 3.0])
+    assert res[7] == ([1, 2, 3], [1, 2, 3])
+    res = client.online_batch_get_features('nycTaxiDemoFeature',
```
Comment (jaymo001) [inline feathr_project/test/test_azure_synapse_e2e.py:80]: I think it should be multi_get_online_features?

File: src/main/scala/com/linkedin/feathr/offline/generation/outputProcessor/PushToRedisOutputProcessor.scala [hunk 6]
```diff
@@ -50,4 +71,213 @@ private[offline] class PushToRedisOutputProcessor(config: OutputProcessorConfig,
       .save()
     (df, header)
   }
+
+  private[feathr] def encodeDataFrame(header: Header, df: DataFrame): DataFrame = {
```
Comment (jaymo001) [inline src/main/scala/com/linkedin/feathr/offline/generation/outputProcessor/PushToRedisOutputProcessor.scala:75]: Looks like this function is not used but duplicated in the Redis push process ?

File: src/main/scala/com/linkedin/feathr/offline/generation/outputProcessor/PushToRedisOutputProcessor.scala [hunk 7]
```diff
@@ -50,4 +71,213 @@ private[offline] class PushToRedisOutputProcessor(config: OutputProcessorConfig,
...
+      }
+    }.toArray
+    StructType(newDfSchemaFields)
+  }
+
+  /**
+   * Gets the function that converts the original data into protobuf data. The index of the schema is fixed so we
+   * map each index to a fixed function.
+   */
+  private[feathr] def getConversionFunction(dfSchema: StructType, allFeatureCols: Map[String, FeatureTypes]): Map[Int, Any => Any] = {
```
Comment (jaymo001) [inline src/main/scala/com/linkedin/feathr/offline/generation/outputProcessor/PushToRedisOutputProcessor.scala:120]: Does allFeatureCols need to be a map[string, featuretypes]?

File: feathr_project/test/test_azure_synapse_e2e.py (hunk 4 above)
Comment (hangfei) [reply feathr_project/test/test_azure_synapse_e2e.py:58]: this is in the test config file.

File: feathr_project/feathr/protobuf/featureValue_pb2.py (hunk 1 above)
Comment (hangfei) [reply feathr_project/feathr/protobuf/featureValue_pb2.py:17]: this is generated.

File: feathr_project/feathr/client.py [hunk 8]
```diff
@@ -217,7 +217,55 @@ def multi_get_online_features(self, feature_table, keys, feature_names):
...
+                    typed_result.append((feature_value.sparseFloatArray.integers, feature_value.sparseFloatArray.floats))
+                elif feature_value.WhichOneof('FeatureValueOneOf') == 'sparseDoubleArray':
+                    typed_result.append((feature_value.sparseDoubleArray.integers, feature_value.sparseDoubleArray.doubles))
+                elif feature_value.WhichOneof('FeatureValueOneOf') == 'sparseLongArray':
+                    typed_result.append((feature_value.sparseLongArray.index_integers, feature_value.sparseLongArray.longs))
+                else:
+                    # fail to load the type. Maybe a new type that is not supported by this client version
+                    typed_result.append(None)
+            else:
+                typed_result.append(raw_feature)
```
Comment (jaymo001) [inline feathr_project/feathr/client.py:267]: This is just  typed_result.append(None), is this expected?

Comment (hangfei) [reply feathr_project/feathr/client.py:267]: yes. see the comment so it fails gracefully

**Review by jaymo001 — APPROVED** [review]

**Review by jaymo001 — DISMISSED** [review]

**Review by xiaoyongzhu — APPROVED** [review]

**Review by jaymo001 — APPROVED** [review]

**Pull request merged**